### ISSUE TITLE
Brave: LAN discovery & announcement  (mDNS, DNS-SD)

### ISSUE THREAD

**Issue by lidel** [ipfs-shipyard/ipfs-companion#758]
> Part of work on Embedded JS-IPFS in Brave #716 :lion: and epic about local discovery in browsers (https://github.com/ipfs/in-web-browsers/issues/45)

## Context

https://github.com/ipfs-shipyard/ipfs-companion/pull/754 introduced TCP transport and mDNS-based discovery of nodes in LAN via [js-libp2p-mdns](https://github.com/libp2p/js-libp2p-mdns). 

`libp2p-mdns` is pretty useful, as it aims to support both existing mDNS discovery service provided by every go-ipfs node (`mdns:compat`) and the new  DNS Service Discovery (DNS-SD) described in  [libp2p/specs/discovery/mdns.md](https://github.com/libp2p/specs/blob/master/discovery/mdns.md) that aims to eventually replace it.

**TL;DR**
In other words, embedded js-ipfs running in Brave (#716) is able  to discover go-ipfs  on the same network and connect to it over regular TCP transport:

> ![](https://user-images.githubusercontent.com/157609/64542113-41a29180-d323-11e9-8f32-3a58ab869748.png)

## Known Issues

- `mdns` discovery is passive at the moment
  - `mdns:compat` runs two services: one that queries, and one that responds to queries. Responder tries to listen on [port  5353](https://github.com/libp2p/js-libp2p-mdns/blob/c6d1d49a20e6dd3d43f33d00951d7f70737a6971/src/compat/constants.js#L6) and fails to start if something is already using the port.
  - `chrome.sockets.udp` does not support opening sockets with `SO_REUSEADDR` or `SO_REUSEPORT` ([bugs.chromium.org#238819](https://bugs.chromium.org/p/chromium/issues/detail?id=238819)
  - this means it is very likely responding to mdns queries won't be enabled for most users, and DNS discovery is mostly passive (js-ipfs can discover other nodes and connect to them, but if we have two Brave nodes on the network those need to discover each other by other means)

## Areas to explore

- fix/tweak existing [mdns spec](https://github.com/libp2p/specs/blob/master/discovery/mdns.md) or create a new discovery module specific to brave/browsers
- https://developer.chrome.com/apps/sockets_udp provides APIs related to multicast groups
  - [manifest](https://developer.chrome.com/apps/manifest/sockets) may require `multicastMembership: ""` – https://github.com/GoogleChrome/chrome-app-samples/issues/119#issuecomment-42977744
- https://developer.chrome.com/apps/mdns seems to be read-only, but worth revisiting

## References

- `SO_REUSEADDR` and `SO_REUSEPORT` in http://man7.org/linux/man-pages/man7/socket.7.html
- Chromium bugs / threads about address reuse in `chrome.sockets.udp`
  - [Issue 246867: socket API should support address reuse](https://bugs.chromium.org/p/chromium/issues/detail?id=246867)
  - [Issue 238819: Add support for address reuse in UDP sockets in platform apps](https://bugs.chromium.org/p/chromium/issues/detail?id=238819)
  -  [groups.google.com/chromium.org: UDP socket API address reuse?](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-apps/mX4ZgsqsSaA)

**Comment by lidel** [comment]
https://github.com/ipfs-shipyard/ipfs-companion/issues/967#issue-789971624
